com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java

```diff
@@ -23,29 +32,119 @@
+		return getInitializedEngineChecked();
+	}
+
+	/**
+	 * Try to initialize the VIATRA engine if it hasn!t been before and if the 
+	 * initialization is not possible, it throws a ViatraQueryException.
+	 * 
+	 * @return the initialized engine
+	 * @throws ViatraQueryException if initialized engine is not available yet
+	 */
+	public AdvancedViatraQueryEngine getInitializedEngineChecked() {
+		return getInitializedEngine()
+				.orElseThrow(() ->
+					new ViatraQueryException(MESSAGE_ENGINE_NOT_READY, MESSAGE_ENGINE_NOT_READY));
+	}
+
+	/**
+	 * Provide the initialized VIATRA if possible.<br />
+	 * If the engine hasn't been initialized but it can be, then it will be initialize and
+	 * every initialization action registered before will be executed.
+	 * 
+	 * @return the initialized VIATRA engine if possible, otherwise an empty Optional
+	 * @see ViatraQueryAdapter#getInitializedEngine(Consumer)
+	 * @see ViatraQueryAdapter#initializeQueryWhenEngineReady(IQuerySpecification)
+	 */
+	public Optional<AdvancedViatraQueryEngine> getInitializedEngine() {
+		if(!isInitialized) {
+			initializeEngine();
+		}
 		return engine;
 	}
 	
+	/**
+	 * Checks if an engine is available and execute the given action on it if yes.
+	 * If not, the action will be registered in the list of actions which need to be done 
+	 * when the engine is ready and will be executed when the engine become available.
+	 * 
+	 * @param action the operation which would like to use the initialized VIATRA engine
+	 * @return the initialized VIATRA engine of it is available, otherwise empty Optional.
+	 */
+	public Optional<AdvancedViatraQueryEngine> getInitializedEngine(Consumer<AdvancedViatraQueryEngine> action) {
+		if(action != null) {
+			initializationActions.add(action);
+		}
+		return getInitializedEngine();
+	}
+	
+	private Optional<AdvancedViatraQueryEngine> initializeEngine() {
+		if(!isInitialized && ProjectUtilities.isLoaded(project.getPrimaryProject())) {
+			engine = Optional.of(createQueryEngine(project, notifiers));
+			notifiers = new Notifier[0];
+			isInitialized = true;
+		}
+		if(!isInitialized) {
+			LOGGER.warn(MESSAGE_ENGINE_NOT_READY);
+		} else {
+			initializationActions.forEach(action -> action.accept(engine.get()));
+			initializationActions.clear();
+		}
+		return engine;
+	}
+	
+	/**
+	 * Create a matcher for the VIATRA query engine of the adapter. If it cannot be created
+	 * immediately, the initialization of the matcher will be done when the engine is ready.
+	 * 
+	 * @param <M> The IQuerySpecification implementation specific type of the 
+	 * ViatraQueryMatcher which would be initialized by this method
+	 * @param querySpecification is the query for which an initialized matcher is needed
+	 * @return the matcher if it could haven been initialized or an empty optional if not
+	 */
+	public <M extends ViatraQueryMatcher<? extends IPatternMatch>> Optional<M> 
+			initializeQueryWhenEngineReady(IQuerySpecification<M> querySpecification) {
```

Comment:
I think the name of this method could simply be `requireMatcher` - together with the Optional return value it expresses that the matcher might not be initialized at the moment of the call.

Reply:
I renamed the method according to your suggestion.